include/wil/safearrays.h

```diff
@@ -0,0 +1,1105 @@
+    //!             // Even if we bail early, nothing will leak
+    //!             RETURN_HR_IF_NULL(E_OUTOFMEMORY, bstr);
+    //!         }
+    //!     }
+    //!     *ppsa = sa.release();
+    //!     return S_OK;
+    //! }
+    //! ~~~~
+    //! // Obtain a SAFEARRAY from an API
+    //! HRESULT ProcessWonderful()
+    //! {
+    //!     wil::unique_safearray sa{};
+    //!
+    //!     // Invoke the API
+    //!     RETURN_IF_FAILED(::GetWonderfulData(sa.put()));
+    //!     // Verify the output is expected
+    //!     RETURN_HR_IF(E_UNEXPECTED, sa.vartype() != VT_BSTR);
+    //!     for (auto& bstr : sa.access_data<BSTR>())
+    //!     {
+    //!         // Use the BSTR, no clean up necessary
+    //!         DoSomethingWithBSTR(bstr);
+    //!     }
+    //!     return S_OK;
+    //! }
+    //! ~~~~
+    //! @tparam storage_t   Storage class that manages the SAFEARRAY pointer
+    //! @tparam err_policy  Represents the error policy for the class (error codes, exceptions, or fail fast; see @ref page_errors)
+    //! @tparam element_t   The underlying datatype that is contained in the safearray OR
+    //!                     it can be void to make the safearray generic (and not as typesafe)
+    template <typename storage_t, typename err_policy = err_exception_policy, typename element_t = void>
+    class safearray_t : public storage_t
+    {
+    private:
+        // SFINAE Helpers to improve readability
+        template<typename T>
+        using is_interface_type = wistd::disjunction<wistd::is_same<T, LPUNKNOWN>,
+                                                        wistd::is_same<T, LPDISPATCH>>;
+        template<typename T>
+        using is_pointer_type = wistd::disjunction<wistd::is_same<T, BSTR>,
+                                                        wistd::is_same<T, LPUNKNOWN>,
+                                                        wistd::is_same<T, LPDISPATCH>>;
+
+        template<typename T>
+        using is_type_not_set = wistd::is_same<T, void>;
+        template<typename T>
+        using is_type_set = wistd::negation<is_type_not_set<T>>;
+        template <typename T>
+        using is_value_type = wistd::conjunction<wistd::negation<is_pointer_type<T>>, is_type_set<T>>;
+
+    public:
+        template<typename T>
+        using safearraydata_t = safearraydata_t<T, err_policy>;
+
+        // Represents this type
+        using unique_type = unique_any_t<safearray_t<storage_t, err_policy, element_t>>;
+
+        typedef typename err_policy::result result;
+        typedef typename storage_t::pointer pointer;
+        typedef element_t elemtype;
+
+    public:
+        // forward all base class constructors...
+        template <typename... args_t>
+        explicit safearray_t(args_t&&... args) WI_NOEXCEPT : storage_t(wistd::forward<args_t>(args)...) {}
+
+        // Exception-based construction
+        template<typename T = element_t, typename wistd::enable_if<is_type_not_set<T>::value, int>::type = 0>

```

Comment:
```suggestion
        template<typename T = element_t, wistd::enable_if_t<is_type_not_set<T>::value, int> = 0>
```
Use the alias template throughout. Less text and much simpler to read